File: Projects/Offensive/Obfuscation/packer_toolkit.c
/**
 * Obfuscation & Packer Research Toolkit - C Implementation
 * Compile: gcc -o packer_toolkit packer_toolkit.c -lz -llzma -lcrypto
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>
#include <zlib.h>
#include <lzma.h>
#include <openssl/evp.h>
#include <openssl/rand.h>

#define MAX_BUFFER_SIZE 1024*1024  // 1MB
#define ENTROPY_THRESHOLD 7.0

typedef struct {
    uint8_t *data;
    size_t size;
} buffer_t;

typedef struct {
    char name[9];
    uint32_t virtual_size;
    uint32_t raw_size;
    uint32_t characteristics;
    double entropy;
} pe_section_t;

// Simple XOR obfuscation
void xor_obfuscate(uint8_t *data, size_t size, uint8_t key) {
    for(size_t i = 0; i < size; i++) {
        data[i] ^= key;
    }
}

// Calculate Shannon entropy
double calculate_entropy(uint8_t *data, size_t size) {
    if(size == 0) return 0.0;
    
    int frequency[256] = {0};
    double entropy = 0.0;
    
    // Calculate byte frequencies
    for(size_t i = 0; i < size; i++) {
        frequency[data[i]]++;
    }
    
    // Calculate entropy
    for(int i = 0; i < 256; i++) {
        if(frequency[i] > 0) {
            double probability = (double)frequency[i] / size;
            entropy -= probability * (__builtin_popcount(frequency[i]) - 1);
        }
    }
    
    return entropy;
}

// Compress using zlib
buffer_t compress_zlib(uint8_t *data, size_t size) {
    buffer_t result = {NULL, 0};
    uLongf compressed_size = compressBound(size);
    result.data = malloc(compressed_size);
    
    if(compress(result.data, &compressed_size, data, size) == Z_OK) {
        result.size = compressed_size;
    } else {
        free(result.data);
        result.data = NULL;
    }
    
    return result;
}

// Compress using LZMA
buffer_t compress_lzma(uint8_t *data, size_t size) {
    buffer_t result = {NULL, 0};
    size_t compressed_size = size * 2; // Conservative estimate
    result.data = malloc(compressed_size);
    
    lzma_stream strm = LZMA_STREAM_INIT;
    lzma_ret ret = lzma_easy_encoder(&strm, 6, LZMA_CHECK_CRC64);
    
    if(ret == LZMA_OK) {
        strm.next_in = data;
        strm.avail_in = size;
        strm.next_out = result.data;
        strm.avail_out = compressed_size;
        
        ret = lzma_code(&strm, LZMA_FINISH);
        if(ret == LZMA_STREAM_END) {
            result.size = compressed_size - strm.avail_out;
        } else {
            free(result.data);
            result.data = NULL;
        }
    }
    
    lzma_end(&strm);
    return result;
}

// AES encryption (simplified)
buffer_t encrypt_aes(uint8_t *data, size_t size, uint8_t *key, uint8_t *iv) {
    buffer_t result = {NULL, 0};
    EVP_CIPHER_CTX *ctx;
    int len;
    int ciphertext_len;
    
    // Create and initialize context
    if(!(ctx = EVP_CIPHER_CTX_new())) {
        return result;
    }
    
    // Initialize encryption
    if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, key, iv)) {
        EVP_CIPHER_CTX_free(ctx);
        return result;
    }
    
    // Allocate buffer for ciphertext
    result.data = malloc(size + EVP_CIPHER_block_size(EVP_aes_256_cbc()));
    
    // Encrypt
    if(1 != EVP_EncryptUpdate(ctx, result.data, &len, data, size)) {
        free(result.data);
        result.data = NULL;
        EVP_CIPHER_CTX_free(ctx);
        return result;
    }
    ciphertext_len = len;
    
    // Finalize encryption
    if(1 != EVP_EncryptFinal_ex(ctx, result.data + len, &len)) {
        free(result.data);
        result.data = NULL;
        EVP_CIPHER_CTX_free(ctx);
        return result;
    }
    ciphertext_len += len;
    
    result.size = ciphertext_len;
    EVP_CIPHER_CTX_free(ctx);
    return result;
}

// Packer detection signatures
typedef struct {
    char *name;
    uint8_t *signature;
    size_t sig_size;
} packer_signature_t;

packer_signature_t signatures[] = {
    {"UPX", (uint8_t*)"UPX!", 4},
    {"ASPack", (uint8_t*)"ASPack", 6},
    {"PECompact", (uint8_t*)"PEC2", 4},
    {"Themida", (uint8_t*)"Themida", 7},
    {NULL, NULL, 0}
};

// Detect packers in file
void detect_packers(uint8_t *data, size_t size) {
    printf("[*] Packer Detection:\n");
    int detected = 0;
    
    for(int i = 0; signatures[i].name != NULL; i++) {
        for(size_t j = 0; j < size - signatures[i].sig_size; j++) {
            if(memcmp(data + j, signatures[i].signature, signatures[i].sig_size) == 0) {
                printf("    [!] Detected: %s\n", signatures[i].name);
                detected = 1;
                break;
            }
        }
    }
    
    if(!detected) {
        printf("    [+] No known packers detected\n");
    }
}

// Analyze PE file structure
void analyze_pe_file(uint8_t *data, size_t size) {
    printf("[*] PE File Analysis:\n");
    
    // Check DOS header
    if(size < 0x40 || data[0] != 'M' || data[1] != 'Z') {
        printf("    [-] Not a valid PE file\n");
        return;
    }
    
    // Get PE header offset
    uint32_t pe_offset = *(uint32_t*)(data + 0x3C);
    if(pe_offset >= size - 0x100) {
        printf("    [-] Invalid PE header offset\n");
        return;
    }
    
    // Check PE signature
    if(memcmp(data + pe_offset, "PE\x00\x00", 4) != 0) {
        printf("    [-] Invalid PE signature\n");
        return;
    }
    
    // Get number of sections
    uint16_t num_sections = *(uint16_t*)(data + pe_offset + 6);
    printf("    [+] Number of sections: %d\n", num_sections);
    
    // Analyze sections
    uint32_t section_offset = pe_offset + 0xF8;
    for(int i = 0; i < num_sections && section_offset < size - 0x28; i++) {
        pe_section_t section;
        memcpy(section.name, data + section_offset, 8);
        section.name[8] = '\0';
        section.virtual_size = *(uint32_t*)(data + section_offset + 8);
        section.raw_size = *(uint32_t*)(data + section_offset + 16);
        
        // Calculate section entropy
        uint32_t section_data_offset = *(uint32_t*)(data + section_offset + 20);
        uint32_t section_data_size = section.raw_size;
        
        if(section_data_offset < size && section_data_size > 0 && 
           section_data_offset + section_data_size <= size) {
            section.entropy = calculate_entropy(data + section_data_offset, section_data_size);
            printf("    Section %s: VirtualSize=0x%X, RawSize=0x%X, Entropy=%.4f\n",
                   section.name, section.virtual_size, section.raw_size, section.entropy);
            
            if(section.entropy > ENTROPY_THRESHOLD) {
                printf("        [!] High entropy - possibly packed/encrypted\n");
            }
        }
        
        section_offset += 0x28;
    }
}

// Pack a file
void pack_file(const char *input_file, const char *output_file, 
               const char *compression, int encrypt) {
    printf("[*] Packing file: %s\n", input_file);
    
    // Read input file
    FILE *f = fopen(input_file, "rb");
    if(!f) {
        printf("[-] Cannot open input file\n");
        return;
    }
    
    fseek(f, 0, SEEK_END);
    size_t file_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    uint8_t *file_data = malloc(file_size);
    fread(file_data, 1, file_size, f);
    fclose(f);
    
    printf("[+] Original size: %zu bytes\n", file_size);
    printf("[+] Original entropy: %.4f\n", calculate_entropy(file_data, file_size));
    
    buffer_t compressed = {NULL, 0};
    
    // Apply compression
    if(strcmp(compression, "zlib") == 0) {
        compressed = compress_zlib(file_data, file_size);
        printf("[+] Zlib compression applied\n");
    } else if(strcmp(compression, "lzma") == 0) {
        compressed = compress_lzma(file_data, file_size);
        printf("[+] LZMA compression applied\n");
    } else {
        compressed.data = malloc(file_size);
        memcpy(compressed.data, file_data, file_size);
        compressed.size = file_size;
        printf("[+] No compression applied\n");
    }
    
    printf("[+] Compressed size: %zu bytes\n", compressed.size);
    printf("[+] Compressed entropy: %.4f\n", calculate_entropy(compressed.data, compressed.size));
    
    // Apply encryption
    if(encrypt) {
        uint8_t key[32], iv[16];
        RAND_bytes(key, sizeof(key));
        RAND_bytes(iv, sizeof(iv));
        
        buffer_t encrypted = encrypt_aes(compressed.data, compressed.size, key, iv);
        if(encrypted.data) {
            free(compressed.data);
            compressed = encrypted;
            printf("[+] AES encryption applied\n");
            printf("[+] Encrypted size: %zu bytes\n", compressed.size);
            printf("[+] Encrypted entropy: %.4f\n", calculate_entropy(compressed.data, compressed.size));
        }
    }
    
    // Write packed file
    f = fopen(output_file, "wb");
    if(f) {
        fwrite(compressed.data, 1, compressed.size, f);
        fclose(f);
        printf("[+] Packed file written: %s\n", output_file);
    } else {
        printf("[-] Cannot write output file\n");
    }
    
    free(compressed.data);
    free(file_data);
}

// Analyze a file
void analyze_file(const char *filename) {
    printf("[*] Analyzing file: %s\n", filename);
    
    FILE *f = fopen(filename, "rb");
    if(!f) {
        printf("[-] Cannot open file\n");
        return;
    }
    
    fseek(f, 0, SEEK_END);
    size_t file_size = ftell(f);
    fseek(f, 0, SEEK_SET);
    
    uint8_t *file_data = malloc(file_size);
    fread(file_data, 1, file_size, f);
    fclose(f);
    
    printf("[+] File size: %zu bytes\n", file_size);
    
    double entropy = calculate_entropy(file_data, file_size);
    printf("[+] Overall entropy: %.4f\n", entropy);
    
    if(entropy > ENTROPY_THRESHOLD) {
        printf("[!] High entropy - likely packed or encrypted\n");
    } else if(entropy > 6.0) {
        printf("[!] Moderate entropy - possibly packed\n");
    } else {
        printf("[+] Low entropy - likely uncompressed\n");
    }
    
    detect_packers(file_data, file_size);
    analyze_pe_file(file_data, file_size);
    
    free(file_data);
}

int main(int argc, char *argv[]) {
    if(argc < 2) {
        printf("Obfuscation & Packer Research Toolkit\n");
        printf("Usage:\n");
        printf("  Pack:   %s pack <input> <output> [zlib|lzma|none] [encrypt]\n", argv[0]);
        printf("  Analyze: %s analyze <filename>\n", argv[0]);
        return 1;
    }
    
    srand(time(NULL));
    
    if(strcmp(argv[1], "pack") == 0 && argc >= 4) {
        const char *compression = (argc > 4) ? argv[4] : "zlib";
        int encrypt = (argc > 5) ? 1 : 1;
        pack_file(argv[2], argv[3], compression, encrypt);
    } else if(strcmp(argv[1], "analyze") == 0 && argc >= 3) {
        analyze_file(argv[2]);
    } else {
        printf("Invalid command or arguments\n");
        return 1;
    }
    
    return 0;
}

File: Projects/Offensive/Obfuscation/packer_toolkit.py
#!/usr/bin/env python3
"""
Obfuscation & Packer Research Toolkit - Python Implementation
Includes packer, analyzer, and detection capabilities
"""

import os
import sys
import struct
import random
import hashlib
import zlib
import base64
import marshal
import lzma
import argparse
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import hashes, padding
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

class CodeObfuscator:
    """Advanced code obfuscation techniques"""
    
    def __init__(self):
        self.obfuscation_methods = [
            'variable_renaming',
            'string_encoding', 
            'control_flow_flattening',
            'dead_code_insertion',
            'instruction_substitution'
        ]
    
    def rename_variables(self, code):
        """Rename variables to meaningless names"""
        variables = self.extract_variables(code)
        mapping = {}
        
        for var in variables:
            new_name = f"var_{random.randint(1000, 9999)}"
            mapping[var] = new_name
        
        for old, new in mapping.items():
            code = code.replace(old, new)
        
        return code
    
    def extract_variables(self, code):
        """Extract variable names from code (simplified)"""
        # This is a simplified version - real implementation would use AST
        variables = set()
        lines = code.split('\n')
        for line in lines:
            if '=' in line and not line.strip().startswith('#'):
                var = line.split('=')[0].strip()
                if var and not var.startswith(' '):
                    variables.add(var)
        return list(variables)
    
    def encode_strings(self, code):
        """Encode strings with various methods"""
        import re
        
        # Find all strings in code
        strings = re.findall(r'\"[^\"]*\"|\'[^\']*\'', code)
        
        for string in strings:
            original = string
            # Choose random encoding method
            method = random.choice(['base64', 'hex', 'xor'])
            
            if method == 'base64':
                encoded = base64.b64encode(string[1:-1].encode()).decode()
                replacement = f"base64.b64decode('{encoded}').decode()"
            elif method == 'hex':
                encoded = string[1:-1].encode().hex()
                replacement = f"bytes.fromhex('{encoded}').decode()"
            else:  # xor
                key = random.randint(1, 255)
                encoded_bytes = [ord(c) ^ key for c in string[1:-1]]
                encoded_hex = ''.join(f'{b:02x}' for b in encoded_bytes)
                replacement = f"''.join(chr(b ^ {key}) for b in bytes.fromhex('{encoded_hex}'))"
            
            code = code.replace(original, replacement)
        
        return code
    
    def flatten_control_flow(self, code):
        """Basic control flow flattening"""
        # This would normally use AST manipulation
        # Simplified version for demonstration
        lines = code.split('\n')
        obfuscated = []
        
        # Add random state machine
        obfuscated.append("state = 0")
        obfuscated.append("while True:")
        obfuscated.append("    if state == 0:")
        
        for i, line in enumerate(lines):
            if line.strip() and not line.strip().startswith('#'):
                indent = "    " * 2
                obfuscated.append(f"{indent}{line}")
                if i < len(lines) - 1:
                    obfuscated.append(f"{indent}state = {i + 1}")
                    obfuscated.append(f"{indent}continue")
                else:
                    obfuscated.append(f"{indent}break")
                obfuscated.append(f"    elif state == {i + 1}:")
        
        obfuscated.append("        break")
        
        return '\n'.join(obfuscated)
    
    def insert_dead_code(self, code):
        """Insert meaningless code that doesn't affect execution"""
        dead_code_patterns = [
            "x = {}; x = None",
            "for _ in range(random.randint(1, 10)): pass", 
            "if random.random() > 1: print('never executed')",
            "__temp = hashlib.md5(str(time.time()).encode()).hexdigest()",
            "while False: break"
        ]
        
        lines = code.split('\n')
        new_lines = []
        
        for line in lines:
            new_lines.append(line)
            if random.random() > 0.7 and line.strip() and not line.strip().startswith('#'):
                dead_code = random.choice(dead_code_patterns)
                new_lines.append(f"    {dead_code}")
        
        return '\n'.join(new_lines)

class BinaryPacker:
    """Advanced binary packing techniques"""
    
    def __init__(self, key=None):
        self.key = key or os.urandom(32)
        self.compression_methods = ['none', 'zlib', 'lzma']
    
    def create_stub(self, packed_data, decryption_key, compression_method):
        """Create unpacking stub"""
        stub_code = f'''
import zlib, lzma, base64
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

class Unpacker:
    def __init__(self):
        self.packed_data = {packed_data}
        self.key = {decryption_key}
    
    def decompress(self, data, method):
        if method == "zlib":
            return zlib.decompress(data)
        elif method == "lzma":
            return lzma.decompress(data)
        return data
    
    def decrypt(self, data, key):
        cipher = Cipher(algorithms.AES(key), modes.CBC(key[:16]))
        decryptor = cipher.decryptor()
        decrypted = decryptor.update(data) + decryptor.finalize()
        # Remove padding
        return decrypted.rstrip(b'\\x00')
    
    def execute(self):
        # Decrypt and decompress
        compressed_data = self.decrypt(self.packed_data, self.key)
        original_code = self.decompress(compressed_data, "{compression_method}")
        exec(original_code)

if __name__ == "__main__":
    unpacker = Unpacker()
    unpacker.execute()
'''
        return stub_code
    
    def pack_file(self, input_file, output_file, compression='zlib', encrypt=True):
        """Pack a Python file with compression and encryption"""
        with open(input_file, 'r') as f:
            original_code = f.read()
        
        # Compress
        if compression == 'zlib':
            compressed_data = zlib.compress(original_code.encode())
        elif compression == 'lzma':
            compressed_data = lzma.compress(original_code.encode())
        else:
            compressed_data = original_code.encode()
        
        # Encrypt
        if encrypt:
            # Simple XOR encryption for demonstration
            key = os.urandom(32)
            encrypted_data = bytearray()
            for i, byte in enumerate(compressed_data):
                encrypted_data.append(byte ^ key[i % len(key)])
            encrypted_data = bytes(encrypted_data)
        else:
            encrypted_data = compressed_data
            key = b'\x00' * 32
        
        # Create packed payload
        packed_payload = base64.b64encode(encrypted_data).decode()
        
        # Generate unpacking stub
        stub = self.create_stub(packed_payload, list(key), compression)
        
        with open(output_file, 'w') as f:
            f.write(stub)
        
        print(f"[+] Packed {input_file} -> {output_file}")
        print(f"[+] Compression: {compression}, Encryption: {encrypt}")
        print(f"[+] Original size: {len(original_code)}, Packed size: {len(stub)}")

class PackerAnalyzer:
    """Analyze packed files and detect packers"""
    
    def __init__(self):
        self.packer_signatures = {
            'UPX': [b'UPX!', b'UPX0', b'UPX1'],
            'ASPack': [b'ASPack'],
            'PECompact': [b'PEC2', b'PEC3'],
            'Themida': [b'Themida'],
            'VMProtect': [b'VMProtect'],
        }
    
    def calculate_entropy(self, data):
        """Calculate Shannon entropy of data"""
        if not data:
            return 0
        
        entropy = 0
        for x in range(256):
            p_x = float(data.count(bytes([x]))) / len(data)
            if p_x > 0:
                entropy += - p_x * (p_x.bit_length() - 1)
        
        return entropy
    
    def detect_packer(self, file_path):
        """Detect packer signatures in file"""
        with open(file_path, 'rb') as f:
            content = f.read()
        
        detected_packers = []
        
        for packer, signatures in self.packer_signatures.items():
            for signature in signatures:
                if signature in content:
                    detected_packers.append(packer)
                    break
        
        return detected_packers
    
    def analyze_file(self, file_path):
        """Comprehensive file analysis"""
        print(f"\n[*] Analyzing: {file_path}")
        print("-" * 50)
        
        # Basic file info
        file_size = os.path.getsize(file_path)
        print(f"[+] File size: {file_size} bytes")
        
        # Read file content
        with open(file_path, 'rb') as f:
            content = f.read()
        
        # Entropy analysis
        entropy = self.calculate_entropy(content)
        print(f"[+] Entropy: {entropy:.4f}")
        
        if entropy > 7.0:
            print("[!] High entropy - likely packed or encrypted")
        elif entropy > 6.0:
            print("[!] Moderate entropy - possibly packed")
        else:
            print("[+] Low entropy - likely uncompressed")
        
        # Packer detection
        packers = self.detect_packer(file_path)
        if packers:
            print(f"[!] Detected packers: {', '.join(packers)}")
        else:
            print("[+] No known packer signatures detected")
        
        # Section analysis (for PE files)
        if content.startswith(b'MZ'):
            self.analyze_pe_file(content)
        
        print("-" * 50)
    
    def analyze_pe_file(self, content):
        """Basic PE file analysis"""
        print("\n[PE File Analysis]")
        
        try:
            # PE header offset
            pe_offset = struct.unpack('<L', content[0x3C:0x40])[0]
            print(f"[+] PE header offset: 0x{pe_offset:X}")
            
            # Number of sections
            num_sections = struct.unpack('<H', content[pe_offset + 6:pe_offset + 8])[0]
            print(f"[+] Number of sections: {num_sections}")
            
            # Analyze sections
            section_offset = pe_offset + 0xF8
            for i in range(num_sections):
                section_name = content[section_offset:section_offset + 8].rstrip(b'\x00')
                section_size = struct.unpack('<L', content[section_offset + 16:section_offset + 20])[0]
                section_entropy = self.calculate_entropy(content[section_offset:section_offset + section_size])
                
                print(f"    Section {section_name.decode('ascii', errors='ignore')}: "
                      f"Size={section_size}, Entropy={section_entropy:.4f}")
                
                section_offset += 0x28
                
        except Exception as e:
            print(f"[-] PE analysis failed: {e}")

def main():
    parser = argparse.ArgumentParser(description='Obfuscation & Packer Research Toolkit')
    subparsers = parser.add_subparsers(dest='command', help='Command to execute')
    
    # Obfuscate command
    obfuscate_parser = subparsers.add_parser('obfuscate', help='Obfuscate Python code')
    obfuscate_parser.add_argument('-i', '--input', required=True, help='Input file')
    obfuscate_parser.add_argument('-o', '--output', required=True, help='Output file')
    obfuscate_parser.add_argument('-m', '--method', choices=['all', 'variables', 'strings', 'controlflow', 'deadcode'],
                                 default='all', help='Obfuscation method')
    
    # Pack command
    pack_parser = subparsers.add_parser('pack', help='Pack Python file')
    pack_parser.add_argument('-i', '--input', required=True, help='Input file')
    pack_parser.add_argument('-o', '--output', required=True, help='Output file')
    pack_parser.add_argument('-c', '--compression', choices=['none', 'zlib', 'lzma'], 
                            default='zlib', help='Compression method')
    pack_parser.add_argument('--no-encrypt', action='store_true', help='Disable encryption')
    
    # Analyze command
    analyze_parser = subparsers.add_parser('analyze', help='Analyze file for packing')
    analyze_parser.add_argument('-i', '--input', required=True, help='Input file')
    
    args = parser.parse_args()
    
    if args.command == 'obfuscate':
        obfuscator = CodeObfuscator()
        with open(args.input, 'r') as f:
            code = f.read()
        
        if args.method == 'all' or args.method == 'variables':
            code = obfuscator.rename_variables(code)
        if args.method == 'all' or args.method == 'strings':
            code = obfuscator.encode_strings(code)
        if args.method == 'all' or args.method == 'controlflow':
            code = obfuscator.flatten_control_flow(code)
        if args.method == 'all' or args.method == 'deadcode':
            code = obfuscator.insert_dead_code(code)
        
        with open(args.output, 'w') as f:
            f.write(code)
        
        print(f"[+] Obfuscated code written to {args.output}")
    
    elif args.command == 'pack':
        packer = BinaryPacker()
        packer.pack_file(args.input, args.output, args.compression, not args.no_encrypt)
    
    elif args.command == 'analyze':
        analyzer = PackerAnalyzer()
        analyzer.analyze_file(args.input)
    
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
